### PR TITLE
Inter-App drag and drop, Toolbar actions for drag and drop

### PR DESCRIPTION
**Pull request by hosy**
## Description
Drag and drop between apps is now possible. As a user you can drag all file types into the app and all file types out of the app.
Furthermore, you can drop items to the toolbar items (move, delete).

## Related Issue
#250 

## Motivation and Context
Better file handling on iPad, better use of multitouch.

## How Has This Been Tested?
- Place the ownCloud app and Apple Files.app in split view. Drag and Drop items between this apps (ownCloud app in file view).
- Drag local items (inside the ownCloud App) to a toolbar item (move, delete)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

____

BUGS & IMPROVEMENTS

- [X] (1) Files can not be dragged https://github.com/owncloud/ios-app/pull/318#issuecomment-475234655
- [X] (2) Drag folders from another app https://github.com/owncloud/ios-app/pull/318#issuecomment-475537609

### CODE REVIEW

**Comment by mneuwert** [comment]
@hosy
- I don't think that dragging to the toolbar-icon is common and well understood UX pattern
- On iPhone somehow also to aim an toolbar icon as drop target is somewhat tedious
- When you hold your finger above the drop target, the symbol in the top right conner of the dragged item is always a green (+) even if the drop target is for example a trash can icon.
- I think it would be nicer to have custom UIDragPreview instead of tableview cell.
- Looks like it is only possible to drag something from the app but not drop files and upload them to ownCloud

**Comment by hosy** [comment]
@mneuwert thank you for your comments. Here are my answers:

> * I don't think that dragging to the toolbar-icon is common and well understood UX pattern
I think this is more a feature for power users. You don't need to know who it works, because you can do it with the common behavior, but you can do it.
> * On iPhone somehow also to aim an toolbar icon as drop target is somewhat tedious
we can try to find a better solution @fschwarz mentioned this example: 
https://twitter.com/tanmays/status/1070219861836607489
> * When you hold your finger above the drop target, the symbol in the top right conner of the dragged item is always a green (+) even if the drop target is for example a trash can icon.
Unfortunately this cannot be changed, it is a standard iOS behavior.
> * I think it would be nicer to have custom UIDragPreview instead of tableview cell.
This was also my idea, but we discussed this here:
https://github.com/owncloud/ios-app/issues/250#issuecomment-463982065
> * Looks like it is only possible to drag something from the app but not drop files and upload them to ownCloud
Should work in both directions. I will check, if it is something wrong. Please don't try on simulator, it is sometimes buggy.

File: ownCloud/Client/ClientQueryViewController.swift [hunk 1]
```diff
@@ -158,13 +157,13 @@ class ClientQueryViewController: UITableViewController, Themeable {
 		self.navigationItem.rightBarButtonItems = [selectBarButton!, uploadBarButton!]
 
 		// Create bar button items for the toolbar
-		deleteMultipleBarButtonItem =  UIBarButtonItem(barButtonSystemItem: .trash, target: self, action: #selector(actOnMultipleItems))
-		deleteMultipleBarButtonItem?.actionIdentifier = DeleteAction.identifier
-		deleteMultipleBarButtonItem?.isEnabled = false
+		if let dropDelegate = self as? UIDropInteractionDelegate {
```
[the remark this answers is not in the document]
Comment (hosy) [reply ownCloud/Client/ClientQueryViewController.swift]: Do you have a better solution for that? If I don't do this, I will get a warning here: `UIBarButtonItem(image: UIImage(named:"trash"), target: self as AnyObject, action: #selector(actOnMultipleItems), dropTarget: dropDelegate, actionIdentifier: DeleteAction.identifier!)`

[the remark this answers is not in the document]
Comment (hosy) [reply ownCloud/Client/ClientQueryViewController.swift]: Fixed

**Comment by jesmrec** [comment]
Starting Qa...

**Comment by jesmrec** [comment]
### (1)

1. Open the app
2. Try to drag a file

Current: Not posible. Only folders can be dragged
Expected: Both files and folders to be dragged and dropped

iOS12

**Comment by hosy** [comment]
Did you try it on  real device?

**Comment by jesmrec** [comment]
yes, in two real devices: iPhoneX and iPadAir (both with iOS12). I had no problem to drag folders, but i was not able to do the same with files.

**Comment by jesmrec** [comment]
### (2)

Dragging a folder from Files App (iCloud) to ownCloud, causes an error. Dragging files works OK

Mar-22-2019 09-35-42

**Comment by jesmrec** [comment]
just to illustrate (1) i add a GIF

it is easy to drag a folder, but when i try to do the same with a file (item looks highlighted when i pressed it), i do not find the way:

Mar-22-2019 11-53-26

**Comment by hosy** [comment]
@jesmrec strange, here it is working. Please can you recheck?

IMG_0057

**Comment by hosy** [comment]
@jesmrec dragging folders is currently not supported by iOS but I can try to forbid handling folders for inter-app drag & drop

**Comment by jesmrec** [comment]
- i will recheck again, taking different device. Which device is the one you used?
- Yes, forbid inter-app folder dragging is OK

**Comment by jesmrec** [comment]
Latest commit fixes the problem of dragging files

**Comment by jesmrec** [comment]
Also, folders are not allowed to be dragged and dropped from other apps.

**Comment by jesmrec** [comment]
After the fixes from my side it is approved. New issue related with error handling: https://github.com/owncloud/ios-app/issues/335, to be addressed separately, as discussed in #109.

Ready to go.

**Pull request merged**